src/librustc/infer/error_reporting/need_type_info.rs

```diff
@@ -362,6 +433,51 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         err
     }
 
+    /// If the `FnSig` for the method call can be found and type arguments are identified as
+    /// needed, suggest annotating the call, otherwise point out the resulting type of the call.
+    fn annotate_method_call(
+        &self,
+        segment: &hir::ptr::P<hir::PathSegment>,
+        e: &Expr,
+        err: &mut DiagnosticBuilder<'_>,
+    ) {
+        if let (Ok(snippet), Some(tables), None) = (
+            self.tcx.sess.source_map().span_to_snippet(segment.ident.span),
+            self.in_progress_tables,
+            &segment.args,
+        ) {
+            let borrow = tables.borrow();
+            let method_defs = borrow.node_method_def_id();
+            if let Some(did) = method_defs.get(e.hir_id) {
+                let generics = self.tcx.generics_of(*did);
+                if !generics.params.is_empty() {
```

Comment:
So I was debating if this condition is sufficient. I decided it probably is, though we could also choose to be more selective.

One thing is that there won't always be a *direct* connection between the type parameters and the return type. e.g. it might be

```rust
fn foo<T: Iterator>() -> T::Item
```

but, then again, it's still true that specifying `T` would help to figure out the `Item` type. 

Another is that there could be many parameters.

I could imagine searching through the substitutions for the call to detect cases (like `collect`) where the return type is *directly* linked to a parameter and limiting to that case, or highlighting which parameter specifies the return type in the case that there are many (e.g., maybe writing "try writing out the return type `X`, as in `foo.bar::<_, _, X>()`"

But it's probably "good enough" the way it is, tbh, those seem like "potential improvements". (And the best phrasing is sort of unclear anyway.)

Reply:
One thing I did wonder is if we want to be careful around "defaults", though they are currently being phased out on functions:


```rust
fn foo<T, U = u64>()...
```

you could imagine not showing `U` in our suggestions. But nobody quite knows what the semantics of said defaults should be so it's probably "ok as is".

Reply:
```
error: defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions.
 --> file12.rs:1:11
  |
1 | fn foo<T, U = u64>() -> T {
  |           ^
  |
  = note: `#[deny(invalid_type_param_default)]` on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #36887 <https://github.com/rust-lang/rust/issues/36887>
```

with `#[allow(invalid_type_param_default)]` we do something more reasonable already IMO:

```
error[E0282]: type annotations needed
 --> file12.rs:8:13
  |
8 |     let _ = foo();
  |         -   ^^^ cannot infer type for `T`
  |         |
  |         consider giving this pattern a type
```

Reply:
Another interesting case:
```rust
trait T {
    type A;
    fn foo(&self) -> Self::A {
        panic!()
    }
}
struct S<X>(std::marker::PhantomData<X>);

impl<X> T for S<X> {
   type A = X;
}

fn main() {
    S(std::marker::PhantomData).foo();
}
```
```
error[E0282]: type annotations needed
  --> file12.rs:16:5
   |
16 |     S(std::marker::PhantomData).foo();
   |     ^--------------------------------
   |     |
   |     this method call resolves to `<Self as T>::A`
   |     cannot infer type for `X`
```